cmd/lotus-chainwatch/processor/miner.go

```diff
@@ -0,0 +1,655 @@
+	seal_cid text not null,
+	seal_rand_epoch bigint not null,
+	constraint miner_sectors_pk
+		primary key (miner_id, sector_id)
+);
+
+create index if not exists miner_sectors_miner_sectorid_index
+	on miner_sectors (miner_id, sector_id);
+
+create table if not exists miner_info
+(
+	miner_id text not null,
+	owner_addr text not null,
+	worker_addr text not null,
+	peer_id text,
+	sector_size text not null,
+	
+	precommit_deposits text not null,
+	locked_funds text not null,
+	next_deadline_process_faults bigint not null,
+	constraint miner_info_pk
+		primary key (miner_id)
+);
+
+/*
+* captures miner-specific power state for any given stateroot
+*/
+create table if not exists miner_power
+(
+	miner_id text not null,
+	state_root text not null,
+	raw_bytes_power text not null,
+	quality_adjusted_power text not null,
+	constraint miner_power_pk
+		primary key (miner_id, state_root)
+);
+
+/* used to tell when a miners sectors (proven-not-yet-expired) changed if the miner_sectors_cid's are different a new sector was added or removed (terminated/expired) */
+create table if not exists miner_sectors_heads
+(
+	miner_id text not null,
+	miner_sectors_cid text not null,
+	
+	state_root text not null,	
+	
+	constraint miner_sectors_heads_pk
+		primary key (miner_id,miner_sectors_cid)
+    
+);
+
+DO $$
+BEGIN
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'miner_sector_event_type') THEN
+        CREATE TYPE miner_sector_event_type AS ENUM
+        (
+			'ADDED','EXTENDED', 'EXPIRED', 'TERMINATED'
+        );
+    END IF;
+END$$;
+
+create table if not exists miner_sector_events
+(
+    miner_id text not null,
+    sector_id bigint not null,
+    state_root text not null,
+    event miner_sector_event_type not null,
+    
+	constraint miner_sector_events_pk
+		primary key (sector_id, event, miner_id, state_root)
+)
+`); err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+type minerActorInfo struct {
+	common actorInfo
+
+	state miner.State
+
+	// tracked by power actor
+	rawPower big.Int
+	qalPower big.Int
+	ssize    uint64
+	psize    uint64
+}
+
+type sectorUpdate struct {
+	terminationEpoch abi.ChainEpoch
+	terminated       bool
+
+	expirationEpoch abi.ChainEpoch
+
+	sectorID abi.SectorNumber
+	minerID  address.Address
+}
+
+func (p *Processor) HandleMinerChanges(ctx context.Context, minerTips ActorTips) error {
+	minerChanges, err := p.processMiners(ctx, minerTips)
+	if err != nil {
+		log.Fatalw("Failed to process miner actors", "error", err)
+	}
+
+	if err := p.persistMiners(ctx, minerChanges); err != nil {
+		log.Fatalw("Failed to persist miner actors", "error", err)
+	}
+
+	if err := p.updateMiners(ctx, minerChanges); err != nil {
+		log.Fatalw("Failed to update miner actors", "error", err)
+	}
+	return nil
+}
+
+func (p *Processor) processMiners(ctx context.Context, minerTips map[types.TipSetKey][]actorInfo) (out []minerActorInfo, err error) {
+	start := time.Now()
+	defer func() {
+		if err == nil {
```

Comment:
This `err` does not get assigned anywhere below and will always be nil.

Reply:
You're using the return err value, but they are set to `nil` on creation and never changed. (The closures below all use `:=` and will create a new instance of `err` inside and then get destroyed at the end of the outermost `for`.)

Reply:
Disregard. I guess my local testing didn't exercise your pattern exactly the same. This appears to work as intended.